### ISSUE TITLE
Enable inverse transform of predictions at inference-time

### ISSUE THREAD

**Issue by asogaard** [icecube/graphnet#97]
Regressing targets that have highly non-Gaussian distributions and/or have numerical scales that, in their natural units, are very far from O(1), is a common problem (e.g. energy and interaction time). This can be solved in (at least) two ways:

1. By having predicting the targeting in its natural units and then — within the loss function and only for the purposes of numerically stable training — transforming both into a space with characteristic scales of O(1). This is what the `transform_prediction_and_target` callable argument does [here](https://github.com/icecube/graphnet/blob/e3a7ec6827fc3f57903d0e71edcfc5c77057203c/src/gnn_reco/components/loss_functions.py#L25).
2. By transforming the target into a space with characteristic scales of O(1); having the model predict _this_ quantity; and then perform the inverse transform of the predictions at inference-time to obtain the actual physical quantity of interest.

Currently, the `transform_target` callable argument [here](https://github.com/icecube/graphnet/blob/e3a7ec6827fc3f57903d0e71edcfc5c77057203c/src/gnn_reco/components/loss_functions.py#L26) does the first half of (2.) above, but in order to enable this second option fully we need:

1. an argument like `inference_transform` to `Model` [here](https://github.com/icecube/graphnet/blob/e3a7ec6827fc3f57903d0e71edcfc5c77057203c/src/gnn_reco/models/model.py#L17) that, if specified is called on the model outputs when an `inference` flag is set.
2. such an inference flag, e.g. using an `.inference()` method to `Model`, analogous to (and interplaying with) the `.train()` and `.eval()` methods of `torch.nn.Module`, cf. [here](https://pytorch.org/docs/stable/generated/torch.nn.Module.html).

**Comment by asogaard** [comment]
Tagging @kaareendrup, as you have expressed interest in, and have already bumped up against, this. Let me know if this is something you'd like to look into.

**Comment by troelspetersen** [comment]
Previously, the robust scalar (subtract mean and scale by 1/Std.) was a good solution (Bjørn and Mads), and was applied as in (2.). However, would it not be nice to do this as suggested in (1.), as it keeps this part internal? I'm thinking that this makes it easier for an unknowing user...

**Comment by asogaard** [comment]
Yes, that has been my exact thinking, @troelspetersen, and the reason we have gone for (1.) so far: To me, at least, it is conceptually simpler that the model returns physical predictions, and that the end-user doesn't have to worry about setting the appropriate `training/inference` flag.

However, option (1.) requires that the transform be differentiable in order for pytorch to be able to back-propagate gradients through it. This means that it has to be made up of simple algebraic expressions and/or pytorch built-in functions (e.g. `torch.log10`). `sklearn.RobustScaler` and similar linear transformations should work nicely in this setting, but more involved transforms, like the `sklearn.QuantileTransformer` explored by @kaareendrup for the interaction time/vertex reconstruction probably won't work. Here option (2.) could work, since it just requires the transform to be invertible and not necessarily differentiable.

The only issue is that if we decide to support both approaches, we will need to use the `.inference()` flag stuff for all models, regardless of whether and how targets/predictions are transformed (even if this flag will do nothing in most cases), in order to avoid problems in inference code when switching between models/tasks. So we will introduce a slight extra load on the end-user by supporting both approaches, but if it provides us added flexibility that @RasmusOrsoe and @kaareendrup have been keen on and allows us to solve more challenging tasks (e.g. vertexing) then I figure it is probably worth it.

But feel free to give your two cents, y'all, on whether this is a direction we should follow! :)

**Comment by kaareendrup** [comment]
I agree with you @asogaard, that while 1) is ideal, having the option to use 2) opens up a lot more doors, and working this way has definitely been critical to what I've been doing lately. While it does put a slightly larger burden on the end-user to support both approaches, hopefully we can implement 2) it in an opt-in way such that it only comes up when necessary, and then I think it's worth it in the end.

I'd like to throw myself at this issue, unless @troelspetersen thinks my time is better spent on some other task closer to what we are doing in the office

**Comment by RasmusOrsoe** [comment]
I consider 2) to be key functionality. If we remove this freedom we tie our hands too much in my opinion.

**Comment by troelspetersen** [comment]
Just to be sure, I agree that 2) is key... especially in the developing phase. 1) is probably for the "final" phase, when making the models generally available.

**Comment by asogaard** [comment]
Hi @kaareendrup,
As the assignee on this issue, I just wanted to check whether this is something you think is realistic to do in the near-ish future? :)  I am planning a study where this would be beneficial, so I am just trying to get an idea of a realistic timeline. But no rush!

**Comment by kaareendrup** [comment]
Hi! I'm back on the task now, and hopefully it shouldn't take that long to do. Hope it works with your schedule :)

**Comment by kaareendrup** [comment]
@asogaard 
I have created a working solution now, and have created a PR to showcase it. Please let me know what you think. :)
